// ng-lazyload-image/esm5/src/lazyload-image.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { LazyLoadImageDirective } from './lazyload-image.directive';
var LazyLoadImageModule = /** @class */ (function () {
    function LazyLoadImageModule() {
    }
    /**
     * @param {?} options
     * @return {?}
     */
    LazyLoadImageModule.forRoot = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return {
            ngModule: LazyLoadImageModule,
            providers: [{ provide: 'options', useValue: options }]
        };
    };
    LazyLoadImageModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [LazyLoadImageDirective],
                    exports: [LazyLoadImageDirective]
                },] }
    ];
    return LazyLoadImageModule;
}());
export { LazyLoadImageModule };
function LazyLoadImageModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    LazyLoadImageModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    LazyLoadImageModule.ctorParameters;
}
