// ngx-img-fallback/esm2015/src/app/modules/ngx-img-fallback/ngx-img-fallback.directive.js
import * as tslib_1 from "tslib";
/**
 * Created by vadimdez on 28/01/2018.
 */
import { Directive, ElementRef, Input, Output, EventEmitter, Renderer2 } from '@angular/core';
let ImgFallbackDirective = class ImgFallbackDirective {
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.loaded = new EventEmitter();
        this.isApplied = false;
        this.ERROR_EVENT_TYPE = 'error';
        this.LOAD_EVENT_TYPE = 'load';
        this.nativeElement = el.nativeElement;
        this.onError = this.onError.bind(this);
        this.onLoad = this.onLoad.bind(this);
        this.addEvents();
    }
    ngOnDestroy() {
        this.removeErrorEvent();
        this.removeOnLoadEvent();
    }
    onError() {
        if (this.nativeElement.getAttribute('src') !== this.imgSrc) {
            this.isApplied = true;
            this.renderer.setAttribute(this.nativeElement, 'src', this.imgSrc);
        }
        else {
            this.removeOnLoadEvent();
        }
    }
    onLoad() {
        this.loaded.emit(this.isApplied);
    }
    removeErrorEvent() {
        if (this.cancelOnError) {
            this.cancelOnError();
        }
    }
    removeOnLoadEvent() {
        if (this.cancelOnLoad) {
            this.cancelOnLoad();
        }
    }
    addEvents() {
        this.cancelOnError = this.renderer.listen(this.nativeElement, this.ERROR_EVENT_TYPE, this.onError);
        this.cancelOnLoad = this.renderer.listen(this.nativeElement, this.LOAD_EVENT_TYPE, this.onLoad);
    }
};
tslib_1.__decorate([
    Input('src-fallback'),
    tslib_1.__metadata("design:type", String)
], ImgFallbackDirective.prototype, "imgSrc", void 0);
tslib_1.__decorate([
    Output('loaded'),
    tslib_1.__metadata("design:type", Object)
], ImgFallbackDirective.prototype, "loaded", void 0);
ImgFallbackDirective = tslib_1.__decorate([
    Directive({
        selector: '[src-fallback]'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], ImgFallbackDirective);
export { ImgFallbackDirective };
