// ng-lazyload-image/esm5/src/scroll-preset/rect.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var Rect = /** @class */ (function () {
    function Rect(left, top, right, bottom) {
        this.left = left;
        this.top = top;
        this.right = right;
        this.bottom = bottom;
    }
    /**
     * @param {?} element
     * @return {?}
     */
    Rect.fromElement = /**
     * @param {?} element
     * @return {?}
     */
    function (element) {
        var _a = element.getBoundingClientRect(), left = _a.left, top = _a.top, right = _a.right, bottom = _a.bottom;
        if (left === 0 && top === 0 && right === 0 && bottom === 0) {
            return Rect.empty;
        }
        else {
            return new Rect(left, top, right, bottom);
        }
    };
    /**
     * @param {?} _window
     * @return {?}
     */
    Rect.fromWindow = /**
     * @param {?} _window
     * @return {?}
     */
    function (_window) {
        return new Rect(0, 0, _window.innerWidth, _window.innerHeight);
    };
    /**
     * @param {?} inflateBy
     * @return {?}
     */
    Rect.prototype.inflate = /**
     * @param {?} inflateBy
     * @return {?}
     */
    function (inflateBy) {
        this.left -= inflateBy;
        this.top -= inflateBy;
        this.right += inflateBy;
        this.bottom += inflateBy;
    };
    /**
     * @param {?} rect
     * @return {?}
     */
    Rect.prototype.intersectsWith = /**
     * @param {?} rect
     * @return {?}
     */
    function (rect) {
        return rect.left < this.right && this.left < rect.right && rect.top < this.bottom && this.top < rect.bottom;
    };
    /**
     * @param {?} rect
     * @return {?}
     */
    Rect.prototype.getIntersectionWith = /**
     * @param {?} rect
     * @return {?}
     */
    function (rect) {
        var /** @type {?} */ left = Math.max(this.left, rect.left);
        var /** @type {?} */ top = Math.max(this.top, rect.top);
        var /** @type {?} */ right = Math.min(this.right, rect.right);
        var /** @type {?} */ bottom = Math.min(this.bottom, rect.bottom);
        if (right >= left && bottom >= top) {
            return new Rect(left, top, right, bottom);
        }
        else {
            return Rect.empty;
        }
    };
    Rect.empty = new Rect(0, 0, 0, 0);
    return Rect;
}());
export { Rect };
function Rect_tsickle_Closure_declarations() {
    /** @type {?} */
    Rect.empty;
    /** @type {?} */
    Rect.prototype.left;
    /** @type {?} */
    Rect.prototype.top;
    /** @type {?} */
    Rect.prototype.right;
    /** @type {?} */
    Rect.prototype.bottom;
}
