// web-social-share/dist/collection/components/web-social-share-target/web-social-share-target.js
import { h, Host } from '@stencil/core';
import { WebSocialShareFacebook } from '../../utils/web-social-share/web-social-share-facebook';
import { WebSocialShareTwitter } from '../../utils/web-social-share/web-social-share-twitter';
import { WebSocialShareEmail } from '../../utils/web-social-share/web-social-share-email';
import { WebSocialShareLinkedin } from '../../utils/web-social-share/web-social-share-linkedin';
import { WebSocialSharePinterest } from '../../utils/web-social-share/web-social-share-pinterest';
import { WebSocialShareReddit } from '../../utils/web-social-share/web-social-share-reddit';
import { WebSocialShareWhatsapp } from '../../utils/web-social-share/web-social-share-whatsapp';
import { WebSocialShareCopy } from '../../utils/web-social-share/web-social-share-copy';
export class WebSocialShareTarget {
    constructor() {
        this.displayNames = false;
    }
    async componentDidLoad() {
        await this.emitSocialShareLoaded();
    }
    emitSocialShareLoaded() {
        return new Promise((resolve) => {
            if (this.share.facebook) {
                this.socialShareLoaded.emit('facebook');
            }
            else if (this.share.twitter) {
                this.socialShareLoaded.emit('twitter');
            }
            else if (this.share.email) {
                this.socialShareLoaded.emit('email');
            }
            else if (this.share.linkedin) {
                this.socialShareLoaded.emit('linkedin');
            }
            else if (this.share.pinterest) {
                this.socialShareLoaded.emit('pinterest');
            }
            else if (this.share.reddit) {
                this.socialShareLoaded.emit('reddit');
            }
            else if (this.share.whatsapp) {
                this.socialShareLoaded.emit('whatsapp');
            }
            else if (this.share.copy) {
                this.socialShareLoaded.emit('copy');
            }
            resolve();
        });
    }
    handleFacebookShare($event) {
        $event.stopPropagation();
        WebSocialShareFacebook.share(this.share.facebook);
        this.selected.emit();
    }
    handleTwitterShare($event) {
        $event.stopPropagation();
        WebSocialShareTwitter.share(this.share.twitter);
        this.selected.emit();
    }
    handleEmailShare($event) {
        $event.stopPropagation();
        WebSocialShareEmail.share(this.share.email);
        this.selected.emit();
    }
    handleLinkedinShare($event) {
        $event.stopPropagation();
        WebSocialShareLinkedin.share(this.share.linkedin);
        this.selected.emit();
    }
    handlePinterestShare($event) {
        $event.stopPropagation();
        WebSocialSharePinterest.share(this.share.pinterest);
        this.selected.emit();
    }
    handleRedditShare($event) {
        $event.stopPropagation();
        WebSocialShareReddit.share(this.share.reddit);
        this.selected.emit();
    }
    handleWhatsappShare($event) {
        $event.stopPropagation();
        WebSocialShareWhatsapp.share(this.share.whatsapp);
        this.selected.emit();
    }
    async handleCopyShare($event) {
        $event.stopPropagation();
        await WebSocialShareCopy.share(this.share.copy);
        this.selected.emit();
    }
    render() {
        return h(Host, { class: {
                'web-social-share-facebook': this.share.facebook !== undefined,
                'web-social-share-twitter': this.share.twitter !== undefined,
                'web-social-share-email': this.share.email !== undefined,
                'web-social-share-linkedin': this.share.linkedin !== undefined,
                'web-social-share-pinterest': this.share.pinterest !== undefined,
                'web-social-share-reddit': this.share.reddit !== undefined,
                'web-social-share-whatsapp': this.share.whatsapp !== undefined,
                'web-social-share-copy': this.share.copy !== undefined
            } }, this.renderButton());
    }
    renderButton() {
        if (this.share.facebook) {
            return (h("button", { onClick: ($event) => this.handleFacebookShare($event), class: 'web-social-share-button web-social-share-button-facebook' },
                this.renderIcon(),
                this.renderName(this.share.facebook, 'Facebook')));
        }
        else if (this.share.twitter) {
            return (h("button", { onClick: ($event) => this.handleTwitterShare($event), class: 'web-social-share-button web-social-share-button-twitter' },
                this.renderIcon(),
                this.renderName(this.share.twitter, 'Twitter')));
        }
        else if (this.share.email) {
            return (h("button", { onClick: ($event) => this.handleEmailShare($event), class: 'web-social-share-button web-social-share-button-email' },
                this.renderIcon(),
                this.renderName(this.share.email, 'Email')));
        }
        else if (this.share.linkedin) {
            return (h("button", { onClick: ($event) => this.handleLinkedinShare($event), class: 'web-social-share-button web-social-share-button-linkedin' },
                this.renderIcon(),
                this.renderName(this.share.linkedin, 'Linkedin')));
        }
        else if (this.share.pinterest) {
            return (h("button", { onClick: ($event) => this.handlePinterestShare($event), class: 'web-social-share-button web-social-share-button-pinterest' },
                this.renderIcon(),
                this.renderName(this.share.pinterest, 'Pinterest')));
        }
        else if (this.share.reddit) {
            return (h("button", { onClick: ($event) => this.handleRedditShare($event), class: 'web-social-share-button web-social-share-button-reddit' },
                this.renderIcon(),
                this.renderName(this.share.reddit, 'Reddit')));
        }
        else if (this.share.whatsapp) {
            return (h("button", { onClick: ($event) => this.handleWhatsappShare($event), class: 'web-social-share-button web-social-share-button-whatsapp' },
                this.renderIcon(),
                this.renderName(this.share.whatsapp, 'WhatsApp')));
        }
        else if (this.share.copy) {
            return (h("button", { onClick: ($event) => this.handleCopyShare($event), class: 'web-social-share-button web-social-share-button-copy' },
                this.renderIcon(),
                this.renderName(this.share.copy, 'Copy')));
        }
        else {
            return (h("div", null));
        }
    }
    renderIcon() {
        return h("div", { class: "web-social-share-button-icon" },
            h("slot", { name: "facebook" }),
            h("slot", { name: "twitter" }),
            h("slot", { name: "email" }),
            h("slot", { name: "linkedin" }),
            h("slot", { name: "pinterest" }),
            h("slot", { name: "reddit" }),
            h("slot", { name: "whatsapp" }),
            h("slot", { name: "copy" }));
    }
    renderName(displayAttributes, defaultBrandName) {
        if (this.displayNames) {
            return (h("p", null, displayAttributes && displayAttributes.brandName && displayAttributes.brandName !== '' ? displayAttributes.brandName : defaultBrandName));
        }
        else {
            return (h("span", null));
        }
    }
    static get is() { return "web-social-share-target"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["web-social-share-target.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["web-social-share-target.css"]
    }; }
    static get properties() { return {
        "displayNames": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "display-names",
            "reflect": false,
            "defaultValue": "false"
        },
        "share": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "WebSocialShareInputConfig",
                "resolved": "WebSocialShareInputConfig",
                "references": {
                    "WebSocialShareInputConfig": {
                        "location": "import",
                        "path": "../../types/web-social-share/web-social-share-input"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        }
    }; }
    static get events() { return [{
            "method": "selected",
            "name": "selected",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": ""
            },
            "complexType": {
                "original": "void",
                "resolved": "void",
                "references": {}
            }
        }, {
            "method": "socialShareLoaded",
            "name": "socialShareLoaded",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": ""
            },
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            }
        }]; }
}
